spark/src/main/scala/com/gu/tableversions/spark/filesystem/PathMapper.scala

```diff
@@ -0,0 +1,91 @@
+package com.gu.tableversions.spark.filesystem
+
+import java.net.URI
+
+import com.gu.tableversions.core.{Partition, Version}
+import org.apache.hadoop.fs.Path
+
+/**
+  * Map paths between one "outer" FileSystem (e.g. a ProxyFileSystem) and an underlying one.
+  * For example, by changing the schema, or appending version directories.
+  */
+trait PathMapper {
+
+  /** Convert a path with a "versioned" scheme to one suitable for the underlying FileSystem. */
+  def forUnderlying(path: Path): Path
+
+  /** Convert a path from the underlying FileSystem to one in the "versioned" scheme. */
+  def fromUnderlying(path: Path): Path
+}
+
+/**
+  * Path mapper that converts paths for partitions according to a pre-configured set of versioned partitions.
+  */
+class VersionedPathMapper(underlyingFsScheme: String, partitionVersions: Map[Partition, Version]) extends PathMapper {
+
+  import VersionedPathMapper._
+
+  override def forUnderlying(path: Path): Path = {
+    assert(path.toUri.getScheme == VersionedFileSystem.scheme,
+           s"Path provided to `forUnderlying` ($path) not in the ${VersionedFileSystem.scheme} scheme")
+
+    appendVersion(setUnderlyingScheme(path))
+  }
+
+  override def fromUnderlying(path: Path): Path = {
+    assert(path.toUri.getScheme == underlyingFsScheme,
+           s"Path provided to `fromUnderlying` ($path) not in the underlying $underlyingFsScheme scheme")
+
+    setVersionedScheme(stripVersion(path))
+  }
+
+  private def stripVersion(path: Path): Path = {
+    partitionVersions
+      .find {
+        case (partition, _) =>
+          path.toString.contains(VersionedPathMapper.normalize(partition.toString))
+      }
+      .map {
+        case (_, version) =>
+          new Path(path.toString.replace(s"/${version.label}", ""))
+      }
+      .getOrElse(path)
+  }
+
+  private def appendVersion(path: Path): Path = {
+    val uri = path.toUri
+    val uriSchemeSpecificPart = uri.getSchemeSpecificPart
+
+    val versionForPath = partitionVersions
```

Comment:
Minor point. `normalize(partition.toString)` is called 3 times. Maybe `map` to `case (k, v) => (normalize(k.toString), v)` or similar